Step4 - 요금 조회 PR요청드립니다.

sangjaeoh opened the pull request:
안녕하세요 😃

4단계 - 요금 조회 PR요청드립니다. 🙇🏻‍♂️

감사합니다 :)

sangjaeoh committed:
docs: 요구사항 작성

sangjaeoh committed:
test: PathAcceptanceTest 지하철 이용요금 추가

sangjaeoh committed:
feat: 거리별 요금 정책 추가

sangjaeoh committed:
feat: 노선에 추가요금 필드 추가

sangjaeoh committed:
feat: 노선별 추가 요금 정책 추가

sangjaeoh committed:
feat: 연령별 할인 정책 추가

sangjaeoh committed:
refactor: Path

sangjaeoh committed:
refactor: FareCalculator

sangjaeoh committed:
fix: resolveArgument

wheejuni commented on src/main/java/nextstep/subway/auth/domain/AuthenticationPrincipal.java:
```diff
@@ -8,4 +8,6 @@
 @Retention(RetentionPolicy.RUNTIME)
 public @interface AuthenticationPrincipal {
+
+    boolean required() default true;
```
언뜻 보았을때는 이 애노테이션이 사용될 파라미터에 직접 붙는 것인데, `required` 여부를 조절할 수 있어야 한다는게 잘 공감이 가지 않습니다 🙇

wheejuni commented on src/main/java/nextstep/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -24,8 +26,16 @@ public boolean supportsParameter(MethodParameter parameter) {
+            WebDataBinderFactory binderFactory) {
         String credentials = AuthorizationExtractor.extract(webRequest.getNativeRequest(HttpServletRequest.class));
+        if (Objects.isNull(credentials) || !isRequired(parameter)) {
+            return new LoginMember();
```
이런 목적이라면 결국 익명 유저의 로그인을 허용하는 것인데 `required` 가 아닌 다른 옵션으로 이름이 바뀌어야 할 거 같군요

sangjaeoh replied:
좀더 명확한 뜻을 가질 수 있도록 `anonymous`로 변경하였습니다!

wheejuni commented on src/main/java/nextstep/subway/path/application/PathService.java:
```diff
@@ -25,14 +26,14 @@ public PathService(StationRepository stationRepository, LineRepository lineRepos
         PathFinder pathFinder = DijkstraShortestPathFinder.from(lines);
         Path path = pathFinder.findPath(sourceStation, targetStation);
-        return PathResponse.from(path);
+        return PathResponse.of(path.getStations(), path.getDistance(), path.calculateFare(Optional.ofNullable(age).orElse(0)));
```
`Path` 에서 미리 요금 계산을 수행한 다음에 (`path.calculateFare()`) `PathResponse.of()` 에는 `Path` 만 넘겨서 변환하는 형태로 설계되는 것이 조금 더 깔끔할 것 같습니다. 지금 형태를 보면 생성자와 크게 차이가 안 나는 거 같아서요

sangjaeoh replied:
`path`의 `calculateFare`의 함수가 값을 반환하도록 하려다 보니 위와 같은 형태가 되었습니다.

피드백 주신대로 `Path`가 `calculateFare`를 하여 미리 계산을 수행하고,
`PathResponse`가 `Path`만 받아 생성하도록 변경하였습니다!

wheejuni commented on src/main/java/nextstep/subway/path/domain/FareAgeCalculator.java:
```diff
@@ -0,0 +1,17 @@
+        return Fare.from(calculateFare(fare, fareAge.getRate()));
+    }
+
+    private static int calculateFare(int fare, double rate) {
+        return (int) ((fare - DISCOUNT_FARE) * rate);
+    }
```
연령대별 할인 요금 산식을 `FareAge` 쪽에 뒀기때문에 계산하는 책임도 그 쪽에 부여해도 된다는 생각입니다.

sangjaeoh replied:
위 부분은 저도 고민을 했었습니다 🤔

`FareAge` 에 할인율도 작성을 해놓아서 `FareAge`에서 계산을 하도록 해야 하나...
아니면 계산에 필요한 값은 `FareAge`에 다 있지만 역할 분리로 계산은 `FareAgeCalculator`에서 해야 하나 말이죠...
`FareAge`에서 계산까지 다 해버리면 `FareAgeCalculator`가 필요 없어 보이기도 하고... 😭

그렇다고 통일감을 위해서 같은 구조인 `FareDistanceCalculator`의 계산의 역할을 `FareDistance`에 넘겨주면
`FareDistance`는 오히려 너무 비대해지는거 아닌가 라는 생각도 들고...

아니면 `FareAge`의 할인율을 `FareAgeCalculator`에 정의할까도 생각들고...
참 어려운거 같습니다 😓

wheejuni reviewed (changes requested):
고생하셨습니다. 💯 
몇 가지 의견 드립니다. 확인 부탁드립니다.

sangjaeoh committed:
refactor: AuthenticationPrincipal 옵션명 변경

sangjaeoh committed:
refactor: Path, PathResponse

- PathResponse가 Path를 받아 생성하도록 변경
- PathResponse의 메소드명 of에서 from으로 변경
- Path의 calculateFare를 void로 변경

sangjaeoh committed:
refactor: FareAgeCalulator

wheejuni reviewed (approved):
이미 많은 고민을 하신 부분인 거 같아 이정도로 충분한 거 같습니다 💯 
수고하셨습니다!
